### DISCUSSION TITLE
Development

### DISCUSSION THREAD

**Discussion by Krrishdhaneja** [Krrishdhaneja/virtual-assistants#10] [General]
@fm-sys would you help me up in this project?

**Comment by fm-sys** [comment]
If there is something specifically, I would help. But please describe in detail what the problem is. Over Christmas I will take a Github break but after that I am looking forward to a challenge 😉

**Reply by Krrishdhaneja** [reply]
this program is in cli(which i think is boring) so i tried to make a gui (using tkinter) but output i was getting was appearing when the program ended(cuz window.mainloop() is at end but if i put it over any of the function then the function wouldn't work). so the main problem here is i want the gui window to come as soon as the program starts. Any suggestions @fm-sys ?

**Reply by Krrishdhaneja** [reply]
🤔

**Reply by fm-sys** [reply]
What do you think about following UI? I will try to implement it in early January...

![Screenshot_20201223_223528_com google android keep](https://user-images.githubusercontent.com/64581222/103038723-7285f480-456f-11eb-90c5-71fa337c7286.png)
So basically one big screen filling textarea and one input line at the bottom. Can you provide me the API keys which I saw in your code?

**Reply by Krrishdhaneja** [reply]
The UI is good but i think there must be a mic. Button if the user wants to speak. And wouldn’t the conversation look good it it was the whatsapp style?

**Reply by Krrishdhaneja** [reply]
Here is what i mean ->
https://drive.google.com/file/d/1TTRLlAJmB_lt7TeuVZABGBJGsVDE2sum/view?usp=drivesdk

**Reply by Krrishdhaneja** [reply]
@fm-sys ?

**Reply by Krrishdhaneja** [reply]
And i made setup for travis and azure pipelines and python package ci, included each and every step which i used to get the correct output but still build was failing, what can i do ?

**Reply by Krrishdhaneja** [reply]
Merry Christmas @fm-sys

**Reply by fm-sys** [reply]
> Merry Christmas @fm-sys

I wish you the same!

**Reply by fm-sys** [reply]
> And i made setup for travis and azure pipelines and python package ci, included each and every step which i used to get the correct output but still build was failing, what can i do ?

Never have worked with azure pipelines, is it really necessary for your project?

**Reply by fm-sys** [reply]
> The UI is good but i think there must be a mic. Button if the user wants to speak. And wouldn’t the conversation look good it it was the whatsapp style?

I wouldn't invest too much in the design at the beginning but making the user input right-aligned would indeed be possible...

**Reply by Krrishdhaneja** [reply]
> > And i made setup for travis and azure pipelines and python package ci, included each and every step which i used to get the correct output but still build was failing, what can i do ?
> 
> Never have worked with azure pipelines, is it really necessary for your project? 

Ya, cuz I only have a Windows machine and not a Linux, and using a VM is boring and in WSL 2 I can't use the tts.

**Reply by Krrishdhaneja** [reply]
> > The UI is good but i think there must be a mic. Button if the user wants to speak. And wouldn’t the conversation look good it it was the whatsapp style?
> 
> I wouldn't invest too much in the design at the beginning but making the user input right-aligned would indeed be possible... 

Ya, firstly I also wanted to make the gui simple and then go
Harder, so that is why i chose tkinter and not Qt(tkinter is also easy, i think that)

**Reply by Krrishdhaneja** [reply]
But one thing I noticed is that, I have had a testing file for friday, which I added to .gitignore but it was still pushed to github, any ideas about this @fm-sys ?

**Reply by Krrishdhaneja** [reply]
Hey, @fm-sys in which Time zone you live? I am pretty sure that when you're asleep am waking and when you're waking am asleep!

**Reply by fm-sys** [reply]
> Hey, @fm-sys in which Time zone you live? I am pretty sure that when you're asleep am waking and when you're waking am asleep!

Quite possibly. My time zone is UTC+2

**Reply by Krrishdhaneja** [reply]
Alright! That is a 3 hours 30 min. Difference

**Reply by Krrishdhaneja** [reply]
> And i made setup for travis and azure pipelines and python package ci, included each and every step which i used to get the correct output but still build was failing, what can i do ?

What about this?

**Reply by Krrishdhaneja** [reply]
> But one thing I noticed is that, I have had a testing file for friday, which I added to .gitignore but it was still pushed to github, any ideas about this @fm-sys ?

And this @fm-sys

**Reply by Krrishdhaneja** [reply]
Hey, and I wanted to implement gTTS instead of pyttsx3, cuz the voice is much better, but it can only say one `speak()` command cuz it saves data to an mp3, and making multiple mp3s is not a good idea at all, so what should I do @fm-sys ?

**Reply by fm-sys** [reply]
> > And i made setup for travis and azure pipelines and python package ci, included each and every step which i used to get the correct output but still build was failing, what can i do ?
> 
> What about this?

No idea

**Reply by fm-sys** [reply]
> > But one thing I noticed is that, I have had a testing file for friday, which I added to .gitignore but it was still pushed to github, any ideas about this @fm-sys ?
> 
> And this @fm-sys

Maybe you had already added the file to the git index before you had modified .gitignore

**Reply by Krrishdhaneja** [reply]
Nah, I haven't commited that, but when I made that I included that file in .gitingnore.

**Reply by Krrishdhaneja** [reply]
Do you have Linux? @fm-sys

**Reply by fm-sys** [reply]
> Do you have Linux? @fm-sys

Ne except a virtual machine

**Reply by fm-sys** [reply]
> 
> 
> Hey, and I wanted to implement gTTS instead of pyttsx3, cuz the voice is much better, but it can only say one `speak()` command cuz it saves data to an mp3, and making multiple mp3s is not a good idea at all, so what should I do @fm-sys ?

```
>>> from gtts import gTTS
>>> tts = gTTS('hello')
>>> tts.save('temp_file.mp3')
```
and then

```
pip install python-vlc

>>> import vlc
>>> p = vlc.MediaPlayer("file:///path/to/temp_file.mp3")
>>> p.play()
```

...or any other library for playing

**Reply by fm-sys** [reply]
> 
> 
> Nah, I haven't commited that, but when I made that I included that file in .gitingnore.

See https://stackoverflow.com/questions/1143796/remove-a-file-from-a-git-repository-without-deleting-it-from-the-local-filesyste for removing files from the repo

**Comment by fm-sys** [comment]
@Krrishdhaneja 

By the way, you do not need to link users every time you write a comment. Project Owners and people who participate in the issue before automatically receive a notification. Linking them every time could be perceived as intrusive. (I saw you where also linking RobinLinus in the Snapdrop Repo in every comment. You can be sure that the Owners most probably will read your comments even without linking them)

**Reply by Krrishdhaneja** [reply]
Would be remebering next time for sure

**Comment by Krrishdhaneja** [comment]
Hey, should I now delete this conversation cuz there is Friday discussions already, what do you think?

**Reply by fm-sys** [reply]
I would keep it. Many interesting things are documented here...

**Reply by Krrishdhaneja** [reply]
Hey, I added the file api_config.txt in the directory cuz it was causing some build issues and updated the README according to that

**Reply by Krrishdhaneja** [reply]
Do you encountered an error like this OSError: No Default Input Device Available , cuz it was only thing that was interrupting the build.

**Reply by fm-sys** [reply]
> Hey, I added the file api_config.txt in the directory cuz it was causing some build issues and updated the README according to that

Unfortunately if I try to commit changes, it is also adding the API keys to this file, so it might happen, that someone uploads his API keys.

**.gitignore does only work if the file is not already contained in the repository**

Please show me the build error, so that we can add a fix for this error without creating an `api_config.txt` in the repository

**Reply by fm-sys** [reply]
> 
> 
> Do you encountered an error like this OSError: No Default Input Device Available , cuz it was only thing that was interrupting the build.

Probably cause the emulator has no microphone

**Reply by Krrishdhaneja** [reply]
Hmm... so, how can I resolve this ?

**Reply by Krrishdhaneja** [reply]
https://drive.google.com/file/d/1XHeN6OIJsFGz0zFtV0x5z-ZhHIZbLdL7/view?usp=drivesdk

**Reply by Krrishdhaneja** [reply]
Image of failing build

**Reply by Krrishdhaneja** [reply]
@fm-sys that is a great step!

**Reply by Krrishdhaneja** [reply]
Um.. @fm-sys you can now squash the commits!

**Reply by fm-sys** [reply]
> 
> 
> Um.. @fm-sys you can now squash the commits!

I noticed that I don't have the necessary rights to directly push to the master branch. But actually I don't think it's a big problem that they are contained in the commit history. Or can you maybe generate new API keys and deactivate the old ones?

**Reply by Krrishdhaneja** [reply]
But can you tell me how to do that?

**Reply by fm-sys** [reply]
> 
> 
> But can you tell me how to do that?

Use these cmd commands in your project directory:

    git reset --soft 08da87ce555794820e1788740ff1d872d44a6e7b
    git commit --amend -m "The commit message you want"
    git push --force

(`08da87ce555794820e1788740ff1d872d44a6e7b` is the commit hash of your initial commit)

The result should look like this. Be aware that your commit history will then only contain one commit, see: https://github.com/fm-sys/Friday/tree/squash_test

**Reply by fm-sys** [reply]
Before you squash, the keys are also contained here: https://github.com/Krrishdhaneja/Friday/blob/master/build/lib/friday.py

and here: https://github.com/Krrishdhaneja/Friday/blob/master/dist/virtual_assistant-1.3-py3.8.egg

**Reply by Krrishdhaneja** [reply]
Um.. hey, the releases would also be deleted?

**Reply by fm-sys** [reply]
Don't know, I would think most likely not. But some of them also contain the API keys, so it would be good to delete these, which contain the API keys...

**Reply by Krrishdhaneja** [reply]
nope they dont cuz i deleted them 1 month back

**Reply by Krrishdhaneja** [reply]
did it!

**Reply by fm-sys** [reply]
> did it!

👏👏👏

**Reply by fm-sys** [reply]
> 
> 
> nope they dont cuz i deleted them 1 month back

Affected Releases: v1.1 v1.2 v1.3 v1.3.1

If you download the zip asset, the keys are still there...

**Reply by Krrishdhaneja** [reply]
> > nope they dont cuz i deleted them 1 month back
> 
> Affected Releases: v1.1 v1.2 v1.3 v1.3.1
> 
> If you download the zip asset, the keys are still there...

Hmm...

**Reply by Krrishdhaneja** [reply]
Happy New Year @fm-sys !

**Reply by fm-sys** [reply]
> 
> 
> > But can you tell me how to do that?
> 
> Use these cmd commands in your project directory:
> 
> ```
> git reset --soft 08da87ce555794820e1788740ff1d872d44a6e7b
> git commit --amend -m "The commit message you want"
> git push --force
> ```
> 
> (`08da87ce555794820e1788740ff1d872d44a6e7b` is the commit hash of your initial commit)
> 
> The result should look like this. Be aware that your commit history will then only contain one commit, see: https://github.com/fm-sys/Friday/tree/squash_test

By the way, this is something you should really do only in really rare cases, where it is really necessary and nothing else is possible. Because all forks are no longer compatible with the repo after that and so on. 

I also had to delete and recreate my fork. Fortunately I noticed it before I started with the GUI...

**Reply by fm-sys** [reply]
If something like this happens with the API keys again, and you notice it directly you should rather edit the last commit (you can do that with the "amend commit" button when committing in PyCharm)

**Reply by Krrishdhaneja** [reply]
> If something like this happens with the API keys again, and you notice it directly you should rather edit the last commit (you can do that with the "amend commit" button when committing in PyCharm)

Thanks!

**Reply by Krrishdhaneja** [reply]
> If something like this happens with the API keys again, and you notice it directly you should rather edit the last commit (you can do that with the "amend commit" button when committing in PyCharm)

Thanks!
